### ISSUE TITLE
implement Serialize/Deserialize traits for RefCell

### ISSUE THREAD

**Issue by agrover** [serde-rs/serde#179]
The struct I want to serialize contains a `Vec<Rc<RefCell<Foo>>>`. So, this is kind of half a feature request if a feature is needed in serde, or if not, then it's a question about what do I need to do? :smile:

**Comment by oli-obk** [comment]
oh wow, serde (de-)serializes `Rc` and `Arc`... I didn't know that...

I guess it's fine as long as we don't do it for `RefCell`, `Cell` or `Mutex`.

My reasoning: if you have a `Vec<Rc<RefCell<Foo>>>` and the `Vec` contains 50 clones of the same `Rc`, serializing and deserializing again will end up giving you 50 different `Rc` allocations. As long as they don't have interior mutability, this is perfectly fine. You cannot observe the difference on stable Rust (refcounts are only available on nightly). The moment you have interior mutability you can modify one `Rc`'s allocation and after a ser->de cycle the behavior changes.

**Comment by erickt** [comment]
@oli-obk: That's a good point, I didn't consider that when I implemented `Rc` and `Arc`. I'm not strictly opposed to us implementing `Cell`, `RefCell`, and `Mutex`, but it might be worth just documenting the behavior that deserializing a `Rc<RefCell<Foo>>` won't result in sharing the interior object.

I've started toying around with a solution in our brand new [Serde RFCs](https://github.com/erickt/serde-rfcs) in a [yet-to-be-pushed-up RFCs](https://github.com/erickt/serde-rfcs/blob/master/text/0001-serializing-pointers.md), but there are still some kinks to work out on how to do proper deserialization.
